### PR TITLE
RainbowDAO-Protocol-ink!-Phase-2.md

### PR DESCRIPTION
**Pull request by RainbowcityKun**
### Project Abstract

It is an application for a follow-up grant. The [initial grant](https://github.com/w3f/Grants-Program/pull/680) covered the first and second milestones:
- [RainbowDAO-Protocol-ink!-Phase1 Milestone 1 delivery report](https://github.com/w3f/Grant-Milestone-Delivery/pull/334)
- [RainbowDAO-Protocol-ink!-Phase1 Milestone 2 delivery report](https://github.com/w3f/Grant-Milestone-Delivery/pull/384)

RainbowDAO Protocol is a multi-chain DAO infrastructure service protocol, providing all kinds of DAO infrastructure facilities and Web3 tools-kits for various DAO organizations on different chains. To better serve all types of DAO organizations, we have built a complete set of DAO infrastructure technical lines focusing on the whole DAO ecosystem, including eight categories of more than 30 independent Web3 tools-kits.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.

- [x] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

### How Did You Hear About our grants program?

- [x] Social Media
- [ ] Hackathon
- [ ] Personal Recommendation
- [ ] Substrate Builders Program
- [ ] Investor/VC
- [ ] Online Search 
- [ ] Other: _______

### CODE REVIEW

**Comment by RainbowcityKun** [comment]
Hello everyone, we haven't received any feedback yet, is there anything we need to change or update?

**Comment by RainbowcityKun** [comment]
> Hey @RainbowcityKun Sorry, we haven't had the chance to review the application yet. Please allow for some more time, we're going to take a look as soon as we can! 🙏

Ok, thank you very much for your feedback, feel free to leave me a message if you have any questions.

**Review by takahser — CHANGES_REQUESTED** [review]
@RainbowcityKun I skimmed your application and noticed that:

- currently, M1 and M2 seem to have the same deliveries with the same specifications. why is that?
- currently, the delivery specs are not really detailed, so it would be good if you could add more details there (e.g. "We will provide front-end files and deployment tutorials for this DAO tool product." => which use cases will be covered here? what cause do they serve? etc.)
- you didn't deliver wireframes along with your partially UI-focused grant application. Pls notice that usually we require grantees to deliver wireframes in advance if they apply for a UI-focused grant. Hence, please remove M1.1, M2.1 and deliver the wireframes as part of the application instead

**Comment by RainbowcityKun** [comment]
Hi, @takasher, thank you so much for your feedback. Here is our detailed explanation as below.

*  First,  for this application, we aim to to develop two different DAO tool products which are DAO Multi-signature Committee System for the first milestone and DAO NFT Bank System for the second milestone.  These two DAO products both include UI design, front-end files and Ink! smart contracts, with Ink! smart contracts as its core. That's why M1 and M2 look similar. 

*  Second, as you mentioned, we have upgraded the delivery specs, supplementing more modifications.

*  Third, as you said, we have edited M1.1, M2.1, and will add the  wireframes of both DAO tools to make them an important part of the UI design.

If you find other new problems, please feel free to leave us a message, thank you again for your hard work.

**Comment by semuelle** [comment]
Hey @RainbowcityKun. Just to clarify, what @takahser meant was that we expect wireframes or mockups to be part of the grant application for UI-oriented projects. So the wireframes should be in this document, not a deliverable.

**Comment by RainbowcityKun** [comment]
> Hey @RainbowcityKun. Just to clarify, what @takahser meant was that we expect wireframes or mockups to be part of the grant application for UI-oriented projects. So the wireframes should be in this document, not a deliverable.

Hi, @semuelle, thank you very much for your clarification and reminder.

In this application, we provide two product framework diagrams, which cover the complete functional framework of these two products. Are these two frame diagrams ok?

If that doesn't work, we'll go further and design more detailed wireframes.


DAO Multi-signature Committee System 
image


DAO NFT Bank System

image

**Comment by RainbowcityKun** [comment]
Our designers will provide some more detailed project information in the next few days, please wait for our application update.

**Comment by takahser** [comment]
@RainbowcityKun thanks for adding these flow diagrams. However, these are not [wireframes](https://www.experienceux.co.uk/faqs/what-is-wireframing/). Feel free to add some wireframes, so we can understand the structural design of the UI you intend to deliver as part of the grant. Note that they don't have to be pixel-perfect. But they should contain all the important ui elements that you intend to implement. Ideally, you'd add 1 wireframe per page/use case.
Hope this helps, lmk if there's anything else unclear.

**Comment by RainbowcityKun** [comment]
> @RainbowcityKun thanks for adding these flow diagrams. However, these are not [wireframes](https://www.experienceux.co.uk/faqs/what-is-wireframing/). Feel free to add some wireframes, so we can understand the structural design of the UI you intend to deliver as part of the grant. Note that they don't have to be pixel-perfect. But they should contain all the important ui elements that you intend to implement. Ideally, you'd add 1 wireframe per page/use case. Hope this helps, lmk if there's anything else unclear.

Received, @takahser  thank you very much for your reminder, in fact, our designers have already started UI design in advance. We'll do this part first and put it in the application file. If you have any other questions, please feel free to leave a message.

**Comment by RainbowcityKun** [comment]
Hi @takahser, we have just updated our application documents, please review again.

This time we directly provided the UI files for the two DAO tools, which our designers have been busy with recently. However, the design files here are not the final version. We will re-modify and optimize these UI designs during the contract development process, and determine the final version during the actual development process.

If you have new questions in the future, please feel free to leave us a message, thank you again for your hard work.

**Comment by RainbowcityKun** [comment]
Hi @takahser, can review again when you are not busy. See what else we need to update. Our technicians have already started writing contracts. We know you have a lot of work to do and thank you for your hard work.

**Comment by takahser** [comment]
@RainbowcityKun thx for adding the wireframes, we'll have a look as soon as possible.

**Comment by RainbowcityKun** [comment]
Hi @takahser, can you give us some new feedback recently? Our technical engineers have already started writing the code, and they are still not sure what needs to be updated. Thanks again for your hard work!

**Comment by takahser** [comment]
@RainbowcityKun sorry for the delay here, we're currently experiencing a large backlog.
In general, please keep in mind that if you decide to already start on the coding, you're doing this at your own risk as it's not guaranteed that you'll get a sufficient number of approvals for your grant application. We'll try to get back to you as soon as we can.

**Comment by RainbowcityKun** [comment]
Hi @takahser,thank you very much for your feedback, which allowed us to clarify the status further. We'll be patiently waiting for your new review. If you have new content that we need to update or change in the future, we will definitely improve it as soon as possible.

At the same time, no matter whether our grant application is finally approved or not, our technical staff will continue to develop Ink. We will further improve the functional design of our products and contribute to the Polkadot ecosystem.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the grant application. I quickly took a look at it. Currently it says the exact same things for both milestones: “We will provide the complete ink! smart contracts of this DAO tool product, covering the main product logic according to the logical architecture diagram, which is also the main work of this milestone.” Could you specify this more concretely? The way it’s currently written, you can basically remove this delivery from your second milestone, because everything should be covered after the delivery of the first milestone.

**Comment by RainbowcityKun** [comment]
Hi, @Noc2  , thank you very much for your reply to let us see further news. Regarding your question, here I will explain in more detail.

* Our RainbowDAO protocol is the infrastructure of DAO. We will develop various types of tools for different types of DAO organizations to serve the different needs of DAO organizations.


* In this Grant application document our goal is to develop two different DAO tool products, namely the DAO Multi-Signature Committee System and the DAO NFT Bank System.

 * We will develop the DAO multi-signature committee system in the first milestone.The DAO multi-signature  committee system is mainly for voting and fund proposal application of the multi-signature committee organized by DAO. It contains two parts: the first is the election process of the DAO multi-signature management committee; the second is the fund proposal application process.

 * We will develop the DAO NFT bank system in the second milestone.DAO NFT Bank System can deposit ERC20 tokens into banks for lock-up, and can obtain NFT rewards with ERC721 attributes. In this way, the circulation of ERC20 tokens can be controlled and the currency price can be increased.

 * Both of these milestones require the independent development of Ink! Smart contracts and corresponding front-end systems. So you'll see similar deliveries, but actually deliver a different product.

Thanks again for your prompt reply, and if you have any other questions, please feel free to leave us a message.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Could update the application accordingly, so it doesn’t create any confusion?

**Comment by RainbowcityKun** [comment]
> Thanks a lot for the update. But the specification still says: “We will provide the complete ink! smart contracts of this DAO tool product, covering the main product logic according to the logical architecture diagram, which is also the main work of this milestone.” in both cases. Also could you reduce the price of your second milestone. The price per developer per week seems significantly higher than for the first milestone.

Thank you very much for your patient reply. In fact, the second milestone has the same workload as the first milestone, but our developers are more skilled and can develop products faster.
In order to reach consensus faster, we have lowered the price of the 2nd milestone, and we want to get into development work quickly. We will definitely build the best DAO infrastructure product!

**Comment by RainbowcityKun** [comment]
> But the specification still says: “We will provide the complete ink! smart contracts of this DAO tool product, covering the main product logic according to the logical architecture diagram, which is also the main work of this milestone.” 

In this regard, we will elaborate the specific contract information of each product in detail, please wait for our further update.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update! Feel free to ping me once it’s ready.

**Comment by RainbowcityKun** [comment]
Hi, @Noc2  , We have now re-updated our two milestone application documents, please review again:

* In our first milestone, we will provide the complete ink! smart contracts for DAO multi-signature committee system to illustrate the main product logic mentioned in the logical architecture diagram. The DAO multi-signature committee system is mainly for voting and fund proposal application, i.e., one is the election process and the other the fund proposal application process.<br/>In this milestone, we will mainly develop 4 Ink! Smart contracts which are : 1. Erc20Factory contract: generate ERC20 tokens in batches. The ERC20 Token it generates has an additional vote counting function, which is used to vote for the governance committee. 2. CommitteeFactory contract: create a new multi-signature governance committee. 3. CommitteeManager contract:  manage the operation of the entire multi-signature committee. 4. Vault contract:  manage the funds of the entire committee.<br/>The above four categories of contracts are not the final version of the contract. We will make adjustments accordingly in the actual development process. However, the core functionality remains unchanged whatever others may be adjusted. 


* We will provide the complete ink! smart contracts for this DAO NFT bank system as our second milestone. DAO NFT Bank System can deposit ERC20 tokens into banks for lock-up, and can obtain NFT rewards with ERC721 attributes. In this way, the circulation of ERC20 tokens can be controlled and the currency price can be increased.<br/>In this second milestone, we plan to mainly develop 5 Ink! Smart contracts as below:<br/>1. CreatBank contract: create a new NFT bank, each NFT bank needs to bind the corresponding ERC20 token;<br/>2. CreatNFTDeposit contract: create a new NFT token bound to the NFT bank. Each different NFT bank needs to bind a different NFT token.<br/>3. Depsit721 contract: NFT deposit contract, pledge ERC20 tokens in the NFT bank, the bank issues a corresponding number of ERC721 type NFT token which is equivalent to the bank's deposit certificate;<br/>4. TokenExchange contract: NFT exchange contract, when the NFT deposit maturity date comes, the NFT can be exchanged back to the corresponding ERC20 token, which is equivalent to getting the deposit back;<br/>5. LoanMarket contract: the trading market of NFT certificates, which is used for the free exchange of NFT transaction certificates, and obtains liquidity funds for users who hold them.<br/>The above contracts are not finalized and we will make adjustments as the prodcut develops. Meanwhile, the core functionality remains the same while others may be adjusted or modified.

If you have other new questions, please feel free to leave us a message, we will update it as soon as possible, thank you again for your hard work!

**Review by Noc2 — APPROVED** [review]
Thanks for the update. I’m happy to accept it and mark it as ready for review.

**Comment by RainbowcityKun** [comment]
Hi, @Noc2 @takahser @semuelle , thank you very much for acknowledging us. As our team leader, I am also very aware of the problems that existed when the first grant was delivered .

For the previous grant deliveries, we did have a lot of issues as it was the first time our technical team had used ink! for development. There were still some unskilled parts in the process indeed.
At the same time we don't have rigorous testing and auditing in-house, which is what we lacked in our previous delivery.

During the development of this new grant, we will spend more time and energy to improve our code quality, and conduct multiple rounds of internal testings to make the code as complete as possible.

This time we won't rush to deliver milestones. We will deliver the code after it has been completely tested internally and there are no obvious issues.

Improving code quality is the most important task of our grant at this moment. We are confident to deliver a satisfactory result this time. 

We are grateful to your Web3 foundation team for your continued help that allows our team to grow rapidly.

**Review by takahser — CHANGES_REQUESTED** [review]
Hey @RainbowcityKun 
Thanks for your [suggestions](https://github.com/w3f/Grants-Program/pull/933#issuecomment-1168763483) on how to improve the quality of future deliveries, it's appreciated. In general, I'm willing to give you another chance. However, please make sure to test the deliveries thoroughly to avoid situations as with the previous deliveries. Otherwise you might run at risk of having the grant discontinued. Either way you'll still need to reach the necessary number of approvals here.
Furthermore, after taking a closer look at your application, I have a few questions, see my comments.

**Comment by RainbowcityKun** [comment]
Hey @takahser Thanks  for your honest advice to us. Our new deliveries this time will definitely avoid the situations that were present in previous deliveries. Our development this time will take longer time and energy, especially with various internal tests. We will submit it only if there is no problem in the internal test. I believe we will come up with a very good work this time. I go on to answer your other questions.

**Comment by RainbowcityKun** [comment]
> > Compared with these existing projects, RainbowDAO Protocol ink! can handle more complex DAO governance and collaboration. Our protocol itself is more extensible, more flexible, more vital, creative and innovative.
> 
> Can you make some examples, which additional functionality/flexibility/innovation etc. your protocol would cover, when compared to the competitors you mentioned above?

Hey @takahser Thanks for your question. Please let me further answer it in detail.

First allow me to explain the positioning of RainbowDAO Protocol. The protocol we have been developing  is a set of DAO tools that provide services to the build-up of DAO infrastructure rather than merely a single-function product. The ultimate goal of the Protocol is to fulfill the various demands of DAO community governance and fund management based on all types of DAOs.

The entire RainbowDAO Protocol consists of a main protocol and a few multi-functional plug-ins. The main protocol is DAO factory module which enables anyone to create a DAO and set up different departments/sub-DAOs to achieve multi-level management. This had been compeleted in our first Grant dev phase. The multi-functional plug-ins are the additional module apart from the DAO factory, such as the DAO multi-signature committee system and DAO NFT banking system. They are both parts of the entire RainbowDAO Protocol while can be functioned independently. This is also the core feature we'll develope in our second Grant dev phase.

In the future as we complete the development of DAO multi-signature committee system and DAO NFT bankin system, we will merge the two into the RainbowDAO factory, providing more choices for all different DAOs. Tiny DAOs which are in no need for complicated features can choose the simplest module to create a DAO via RainbowDAO Protocol. Some medium-sized DAOs can adopt the DAO multi-signature management committee system and establish their own multi-signature fund management system through this system. And some large-scale DAOs can also use the DAO NFT banking system to manage their own governance tokens.

The ultimate goal of the Rainbow DAO Protocol is to develop DAO tools with different categories and functions, which can be flexibly selected by various types of DAOs. Which functions these DAOs use depends entirely on the actual needs of each DAO. In the end, we only provide DAO infrastructure products for DAO organizations as a provider of DAO tools. This is what we call the basic flexibility, scalability and innovation of the Rainbow DAO protocol.

In this new grant application, we plan to independently develop the DAO multi-signature management committee system and the DAO NFT banking system. For the first milestone we will complete the DAO multi-signature management committee system and for the second milestone, the DAO NFT banking system will be completed. Both of them can exist as completely independent DAO tools, and ultimately meet the diverse needs of community governance and fund management functions of different types of DAOs.

**Comment by takahser** [comment]
@RainbowcityKun thanks for your answers.
I still have some unclarities and hence, I have a few follow-up questions:

> When users lock up the position, they will get an ERC721 NFT reward, i.e., their deposit certificate.

What's the idea behind minting an ERC721 token for locking up ERC20 tokens? Why not mint fungible tokens instead? No matter if it's an NFT or fungible token, it can be traded on secondary markets, so there is some form of liquidity, right?

> bind this token to the DAO NFT banking system to achieve the normal operation of the system, which is the most basic implementation of our system.

What do you mean by that?  What does "bind this token to the DAO NFT banking system" mean in technical terms? What does "normal operation" include?

> However, we will not develop an independent cross-chain bridge in the early stage to solve the problem of cross-chain interaction.

In that case, I'm not sure how you plan to support ERC20/ERC721 tokens. I'm not aware of these standards being existent in ink! at this time. How do you plan to integrate ERC20/ERC721 tokens into your ink! smart contracts? Are you planning to write your own ERC20 and ERC721 implementations on _ink!_?

File: applications/RainbowDAO-Protocol-ink!-Phase-2.md
Comment (takahser) [reply applications/RainbowDAO-Protocol-ink!-Phase-2.md:48]: For reference, this is the answer: https://github.com/w3f/Grants-Program/pull/933#issuecomment-1170907436

**Comment by RainbowcityKun** [comment]
>  In that case, I'm not sure how you plan to support ERC20/ERC721 tokens. I'm not aware of these standards being existent in ink! at this time. How do you plan to integrate ERC20/ERC721 tokens into your ink! smart contracts? Are you planning to write your own ERC20 and ERC721 implementations on ink!?




Hi, @takahser   I'm sorry, I'm on vacation these days and not at work, so the reply is late.Thank you for your these new questions. The constant communication with you helps us sort out our project logics clearer, which is a great improvement as well. Now let me continue answering your questions.

As you said, we are planning to write our own ERC20 and ERC721 implementations on Ink!. The difference is that ERC20 is designed to be the standard token while ERC721 is not. Please allow me to further explain. 

RainbowDAO Protocol's main users are all types of DAOs. Anyone can create a DAO with just an idea, of which can be a media typed DAO, an investment DAO or even an artistic DAO. These users are not professional blockchain developers, with zero knowledge about code or any smart contract languages like Solidity/Ink!. RainbowDAO is designed to provide services for these people who have no programming knowledge.

Hence, we integrated ERC20 factory feature into our entire protocol. Anyone or DAO can create their own ERC20 tokens through certain initial settings with no need for coding. We have implemented this function in the first grant.

Below is the code link of ERC20:

https://github.com/RainbowcityFoundation/RainbowDAO-Protocol-Ink-milestone_1/tree/main/erc20

This is the code link of ERC20 factory:

https://github.com/RainbowcityFoundation/RainbowDAO-Protocol-Ink-milestone_1/tree/main/erc20_factory

Meantime we found that some new teams in the Polkadot ecosystem are also writing standards for Ink! smart contracts, such as Openbrush. In the futre, we will refer to their work to improve the overall quality of our code.

https://github.com/Supercolony-net/openbrush-contracts

**Comment by RainbowcityKun** [comment]
>  What's the idea behind minting an ERC721 token for locking up ERC20 tokens? Why not mint fungible tokens instead? No matter if it's an NFT or fungible token, it can be traded on secondary markets, so there is some form of liquidity, right?

Please let me explain the idea behind minting an ERC721 token.


In our protocol, each DAO will create an ERC20 token as the governance token of the DAO. This governance token is distributed to community members through different methods such as sales, airdrops or contributions, and ultimately all governance token holders manage the entire DAO. This governance token can be freely traded and circulated through decentralized exchanges.

However, as the issuance of ERC20 governance tokens expands and the circulation increases dramatically, if there are not enough new users to purchase these governance tokens, the price of the governance tokens will drop significantly. It is possible that this DAO eventually fails.

To deal with the situation, we developed this DAO NFT banking system. Any community member of DAOs can deposit its governance tokens into this NFT bank, and he can choose to deposit them for 6 months, 1 year, or 2 years, so that his ERC20 tokens are locked in the bank. As more and more people deposit governance tokens into banks, the secondary market circulation of this ERC20 governance token will become less and less, and its sales pressure will decrease, thus its price will increase.

Community users deposit different amounts of ERC20 tokens into the NFT bank, and in return, they obtain an NFT token as a deposit certificate. At maturity, they can use this NFT token to withdraw those ERC20 tokens from the bank.

The NFT tokens we reward/give out here are not standard ERC721 tokens, but embedded with some special information such as the exact amount of ERC20 tokens contained in the NFT and their maturity date. Each NFT is unique and contains different meanings. We call this NFT a financial NFT. It is fundamentally different from the artistic NFTs on the market today. Because each financial NFT contains a different number of ERC20 governance tokens, all of which are supported by real value. This is also an important innovation of our product.

This picture shows the financial attributes contained in the NFT of each reward. The parameters contained in each NFT are different, and each one is unique. This is why we lock ERC20 tokens into banks to get NFTs.


image

**Comment by RainbowcityKun** [comment]
>  What do you mean by that? What does "bind this token to the DAO NFT banking system" mean in technical terms? What does "normal operation" include?

Sorry for the misunderstanding. Please allow me to further explain the product logics.

Our DAO NFT banking system serves different DAOs, and each DAO can create its own NFT bank on our system. The NFT banks created by different DAO organizations are completely different, which need to be initialized.

When a person wants to initialize an NFT bank, he needs to determine which ERC20 token the bank deposits, and when the user deposits this ERC20 token, what is the name of the NFT he gets. Only when this parameter is set, the NFT bank can operate normally.

Hence, our system supports different DAOs to create different NFT banks. The ERC20 tokens deposited by each NFT bank are different, and the rewarded NFTs are also different. A user cannot deposit an ERC20 token into different NFT banks.

This picture shows the basic required information when initializing an NFT bank:

image

**Comment by RainbowcityKun** [comment]
>  Just a comment on that, you could potentially reuse the existing snowbridge to integrate with Ethereum.

Thank you for the suggestion. We will study the principles of Snowbridge and see if we can integrate it into our protocol some day.

At present, our solution for the Ethereum network is that we are developing the Solidity version of the RainbowDAO protocol synchronously. In the future, we will deploy it on Ethereum to serve the DAOs. Here is the code link to the solidity version we have developed:

https://github.com/RainbowDAO/01-RainbowDAO-Factory


Hi, @takahser  Thanks again for your valuable questions that help us well organize our thinking and logics. Please let me know if there is any new questions. We will reply them as soon as possible. We also hope to improve and upgrade our product during the constant communication.

Thanks again to all the attention and support from each friend in the Web3 Foundation. We will continue to focus on the product development and testing to deliver the real high quality work.

**Review by Noc2 — DISMISSED** [review]
Thanks for the very detailed reply here. I would be willing to give you another chance.

**Comment by takahser** [comment]
@RainbowcityKun thanks for your answers, it's becoming clearer.
However, I still have some follow-up questions:

> The NFT tokens we reward/give out here are not standard ERC721 tokens, but embedded with some special information such as the exact amount of ERC20 tokens contained in the NFT and their maturity date. 

1. How are you planning to embed this "special information" - will you just add it to the metadata or will they actually be stored in the on-chain state?
2. What's the incentive to lock up these ERC20 tokens for a defined time-period? The "financial NFTs" act like a bond, right? The fund owner lends the funds in exchange for a certificate/bond/financial NFT. Is my understanding correct?
3. Will any rewards be paid out to owners of these "financial NFTs"? If so, how (i.e. under what conditions) are they minted?

**Comment by RainbowcityKun** [comment]
> Thanks for the very detailed reply here. I would be willing to give you another chance.

Hi, @Noc2, Thank you for your recognition and support. Later we will definitely deliver high quality code work, a wonderful product.



Hi  @takaher,  Thanks for these follow-up questions. They are very good questions that are precise and fully on spot, which truly stimulate us to organize and sharpen our product logics.



> 1.How are you planning to embed this "special information" - will you just add it to the metadata or will they actually be stored in the on-chain state?


Some of our subsequent smart contracts will estimate the value contained in NFTs by reading the number of ERC20 tokens contained in each NFT. For example, In the NFT trade market, the market value of the NFT will be determined by the number of ERC20 tokens contained in the NFT. Hence, these special information will be stored on chain as the NFT's basic attribute.


> 2.What's the incentive to lock up these ERC20 tokens for a defined time-period? The "financial NFTs" act like a bond, right? The fund owner lends the funds in exchange for a certificate/bond/financial NFT. Is my understanding correct?


Yes, your understanding is correct.  Here financial NFTs and bond are similar. The community users will deposit their DAO governance tokens into the NFT bank, in exchange for the certificate/receipt. At maturity, users can it to withdraw the corresponding ERC20 governance tokens from the bank.


> 3.Will any rewards be paid out to owners of these "financial NFTs"? If so, how (i.e. under what conditions) are they minted?


As you said, in order to encourage more users to mint financial NFTs,  more rewards/incentives should be given out to these financial NFT holders. In our original design, we did not mention the giveout ways, instead, we just thought these could be done by DAOs based on their needs.

Thanks to your suggestion/question and to better our product logics, we will add a financial NFTs airdrop feature: any DAO  can screen NFT holders who meet certain conditions and airdrop a certain number of ERC20 tokens to these NFT holders. This feature will be included in our second milestone.

Thanks again for your valuable advice that helps us constantly improve our product logics.

**Review by Noc2 — APPROVED** [review]

**Comment by Noc2** [comment]
@RainbowcityKun Thanks again for your application. I’m sorry to tell you that the rest of the committee isn’t willing to accept your application. Part of it is because of the quality of your previous deliveries. I’m therefore closing the application. Nevertheless, I wish you all the best with the project and you might want to consider alternative funding resources: https://github.com/w3f/Grants-Program#rocket-alternative-funding-sources Let me know if you have any questions.

**Comment by RainbowcityKun** [comment]
Hi, @Noc2  very sorry to hear this news. But in any case, thank you very much for your help and support as always. In the future, we will still build the Web3 world in our own way. And always bless your team.

**Comment by Noc2** [comment]
Thanks a lot!

**Pull request closed without merging**
